### PR TITLE
{WIP} Introduce size_t as a basic type

### PR DESCRIPTION
**Pull request by UplinkCoder**

### CODE REVIEW

**Comment by rikkimax** [comment]
This will need a ptrdiff_t version too.

**Comment by Geod24** [comment]
Why ?

**Comment by UplinkCoder** [comment]
@Geod24 for corrrect header generation in this case.
There are other reasons for wanting to know if someting is actually size_t

**Comment by UplinkCoder** [comment]
@rikkimax ptrdiff_t is much less used than size_t.

**Comment by wilzbach** [comment]
> @rikkimax ptrdiff_t is much less used than size_t.

Yeah, but it's still used.

**Comment by Geod24** [comment]
> @Geod24 for corrrect header generation in this case.

Can you provide a test case ? Because if the problem presents itself with `size_t`, it's likely to present itself with other aliases.

**Comment by UplinkCoder** [comment]
@Geod24 As soon as this is not a WIP anymore I am happy to provide a testcase.
@TurkeyMan you might be interested in this.

**Comment by TurkeyMan** [comment]
What other aliases?

Without this PR, it is impossible to emit correct .di or .h files.
There are other cases where this knowledge is important too.

**Comment by Geod24** [comment]
> What other aliases?
>
> Without this PR, it is impossible to emit correct .di or .h files.
> There are other cases where this knowledge is important too.

That's why I'm asking for a test case. I suppose the issue is that `size_t foo()` will emit `uint64_t foo()` on 64 bits ?

If it is so, adding `size_t` to the compiler will just mask the problem. Anyone using an alias expect that alias to show up in the code. Aliases are declared as `typedef` after all.
At the moment there's a whole lot of features which don't work (e.g. namespaces). When you add support for more features (e.g. `version` for `ifdef`) you'll have that issue crippling back in.

The root cause is that `alias` were originally designed to be completely transparent. No name of their own, no storage classes of their own, just a compiler-known `#define`. Turns out people expect to be able to stick a `deprecated` on them, to give them a visibility that might differ from the underlying symbol, and to see the alias name in error message instead of the underlying symbol.

So yeah, we need to make alias less transparent in the compiler, because many places just work around the behavior of `toAlias`.
Now it's quite a task to tackle, and I understand that you don't want to be blocked on it to get `size_t` to work. In which case, you can just use the same hack that I used when faced with the same problem, and use the pre-semantic AST ([`Declaration.originalType`](https://github.com/dlang/dmd/blob/ca55eb392590ac9797a1b4321f6e817fe82d3bc6/src/dmd/declaration.d#L287-L290)) which is dirty, but works with all functions / alias.

Now that I look at it, I see that:
```D
alias SIZE = size_t;

extern(C++)
{
    struct FooBar (T) { T val; }
    T foobar (T) (FooBar!T value) { return value.val; }
}

alias I1 = foobar!SIZE;
alias I2 = foobar!int;
```

Results in:
```C++
// Automatically generated by Digital Mars D Compiler v2090

#pragma once

#include <assert.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#if !defined(_d_real)
# define _d_real long double
#endif


// Parsing module foo
typedef uint64_t SIZE;

template <typename T>
struct FooBar
{
    // Ignoring var val alignment 0
    T val;
};

// ignored function foo.foobar!ulong.foobar
// ignored function foo.foobar!int.foobar
```

And if you want to support any kind of templated code, you'll need to support arbitrary aliases, because the root cause is the same. That's why cppmangle has so much complication when it comes to templates. Remove eager `toAlias` resolution and you can remove hundreds of lines of code.

**Comment by Geod24** [comment]
Ah I see you submitted an issue @TurkeyMan : https://issues.dlang.org/show_bug.cgi?id=20600

**Comment by TurkeyMan** [comment]
C++ headers are new and immature; you might as well use .di for your thought experiments for the time being.

You're also correct about things like deprecating old names, etc. That's the most common case where this issue comes up in my experience.

**Comment by UplinkCoder** [comment]
This passing is this PR is dependent of the druntime `size_t` alias being removed!
@WalterBright 
@Geod24
changing `alias` to be anything else than a transparent alias is likely to be extremely disruptive.

The druntime alias of size_t is a lie!
If one were to change only that without corresponding compiler changes, really strange things are going to happen.
Rather than lying and creating a further coupling between runtime and compiler I'd rather make `size_t` a built-in type.

**Comment by WalterBright** [comment]
Why do this?

**Comment by UplinkCoder** [comment]
@WalterBright The initial motivation was c++ header generation which is currently done after semantic3() it seems.
at which point there is no way to recover size_t if it is not made a type which can be `distinguished` from `uint` or `ulong` (or `ushort` for 16bit which I'd like to support again one day)

File: src/dmd/root/object.h [hunk 1]
```diff
@@ -35,6 +35,15 @@ enum DYNCAST
...
+#ifdef DEBUG
+    unsigned long long serial;
+    static unsigned long long next_serial;
+
+    RootObject()
+    {
+        serial = next_serial++;
+    }
+#endif
     RootObject() { }
```
Comment (thewilsonator) [inline src/dmd/root/object.h]:
```
#else
    RootObject() { }
#endif
```

File: src/dmd/dcast.d [hunk 2]
```diff
@@ -3319,11 +3319,42 @@ Lagain:
...
             t1 = e1.type;
             t2 = e2.type;
-            goto Lagain;
+
+            if (t1old != t1 || t2old != t2)
+            {
+                // only try this again if the integral promotion changed anything!
+                goto Lagain;
+            }
+            else
```
Comment (thewilsonator) [inline src/dmd/dcast.d]: This doesn't need to be an `else`

Comment (UplinkCoder) [reply src/dmd/dcast.d]: I am afraid if this is not an else, I cannot isolate the only where integral promotion does not change things ...

**Comment by WalterBright** [comment]
> at which point there is no way to recover size_t

Ok, I see. But there is no end to this. This applies to *all* aliases for types.

It's come up before. In the definition of `Declaration`, there's a field `originalType` which is there for this purpose.

**Comment by UplinkCoder** [comment]
@WalterBright I fail to see how originalType could even represent size_t ...

**Comment by WalterBright** [comment]
Besides, I bet you'd get a lot better results by doing the C++ header emission *before* semantic analysis, not after. Note that the `.di` header generation is specifically done before header generation for the reasons you've run into. After semantic, you'll get the reduced and lowered code.

**Comment by WalterBright** [comment]
> I fail to see how originalType could even represent size_t ...

As a `TypeIdentifier`. Grep for it in the sources and see how it is used.

**Comment by UplinkCoder** [comment]
@WalterBright If function definitions are generated by mixins you would not get those in the header.
if it's not run after semantic3()
The `.di` generation has similar problems.

**Comment by TurkeyMan** [comment]
generating functions in mixins is extremely important to me... :/

**Comment by UplinkCoder** [comment]
I am giving the `TypeIdentifier` based recovery a shot.
Closing for now.

Thanks @WalterBright

**Comment by WalterBright** [comment]
> generating functions in mixins is extremely important to me... :/

@TurkeyMan You're always going to have one problem or the other if you want semantic processing here but not there. Note that `size_t` only matters if you want the generated code to be portable 32<=>64. But, if you re-generate the header each time you compile, then `size_t` in the output is not necessary.

Just like nobody tries to generate an object file and use it for both 32 and 64 bit links.

**Comment by WalterBright** [comment]
@TurkeyMan another possibility is for you to declare `size_t` to be a struct for header-only builds. Then it'll show up in the `.h` file as `size_t`.

**Comment by UplinkCoder** [comment]
If it can be recoverd via original type it's okay.

**Comment by TurkeyMan** [comment]
We distribute packages which include suites of binaries for different targets. I don't really want to emit as many bundles of headers as there are binaries, just because of size_t... that feels very lame.

I don't declare size_t, other people do. It's used systemically throughout the language and libraries, and I can't change that.

> Just like nobody tries to generate an object file and use it for both 32 and 64 bit links.

That is not comparable... `size_t` is deeply fundamental, and knowledge of its type should be knowable. An extern statement should know if size_t was used or not.

**Comment by ibuclaw** [comment]
I think you're inventing a non-problem here.  We _always_ emit correct `.h` files even in the absence of `size_t`.

**Comment by UplinkCoder** [comment]
@ibuclaw have you used the experimental `-HC` switch?

**Comment by ibuclaw** [comment]
@UplinkCoder if you want to emit `size_t` to header, then do comparisons with `Type.tsize_t` or `Type.tptrdiff_t`.

**Comment by UplinkCoder** [comment]
> @UplinkCoder if you want to emit `size_t` to header, then do comparisons with `Type.tsize_t` or `Type.tptrdiff_t`.

Definition of Type.tsize_t is: `size_t    = basic[isLP64 ? Tuns64 : Tuns32];`

**Comment by UplinkCoder** [comment]
@WalterBright in case of 
`alias F = extern(C) size_t function (size_t z);`
The `AliasDeclaration.originalType` seems to contain the return type of the function rather than function type itself.

**Comment by UplinkCoder** [comment]
Reopening because utilising `originalType` did not work, it's `null` in certain cases.

**Comment by wilzbach** [comment]
What are these cases? What about fixing it for them?

**Comment by UplinkCoder** [comment]
in the case I posted.
`alias F = extern(C) size_t function (size_t z);` has no way to print the correct output
which is `typedef size_t (*F) (size_t z) F;`

**Comment by TurkeyMan** [comment]
@ibuclaw A non-problem? Header files which replace `size_t` with `int` aren't correct header files.
Have you ever seen a distribution in your life with multiple include trees because size_t was defined differently in one folder compared to the next? Our software distribution systems don't have any expression for multiple include trees selected by different configurations; we would need to extend out distribution solution to hook that up, or find some awkward work-arounds, which is probably possible, but it's a pretty lame issue to solve.

Now, if this was something other than `size_t`, I might find that acceptable, but this is `size_t`, which is SO FUNDAMENTAL, it appears *everywhere*, and I'd wager it's similarly frequent to `int`, and much more common than `short`, and possibly even `float`.
Every array length, every container/collection type, every file/IO interaction, every memory allocator, etc... you can't 'work-around' `size_t`, and we shouldn't have to.

**Comment by TurkeyMan** [comment]
FWIW, it's also not the first time I've had a problem of this nature; I've been having trouble identifying `size_t` for as long as I've been doing reflection in D. In the past, the situation wasn't critical enough that I wasn't satisfied with a workaround, but a long history of past issues with this same problem should probably be taken to add weight to its importance.
It comes up relatively often that reflection code wants to know that `size_t` was actually `size_t`, and any solution here should fix that history of issues too.

**Comment by TurkeyMan** [comment]
I searched DMD, phobos, and druntime for frequency of primitive types:

type | occurances
--- | ---
int | 30000
uint | 8400
size_t | 8000
long | 2200
float | 2000
ulong | 1700
short | 1000

`size_t` is the 3rd most frequently used type, and only just narrowly beaten by `uint`.
Probability of bumping into `size_t` whenever doing reflection based code is extremely high... and in this case, where I might deploy a workaround for a user-defined alias, or something local to a project or module, I don't think the expectation of reasonable when you apply it to `size_t`.

Walter makes the argument that Stefan's solution opens a floodgate, where the problem applies to all aliases, but I make the counter-argument that `size_t` is distinctly special and worthy of special treatment in this case. It's far too important, and reliably affects all code everywhere.

**Comment by Geod24** [comment]
> This would involve creating an AliasDeclaration.originalType at parse time.

It's just a matter of calling `syntaxCopy` when semantic starts.
https://github.com/dlang/dmd/search?q=originalType+path%3Asrc&unscoped_q=originalType+path%3Asrc

**Comment by UplinkCoder** [comment]
LLVM has a call in their API Named: `LLVMIntPtrTypeInContext`
They seem to think `size_t` matters enough to special case.

**Comment by WalterBright** [comment]
> Header files which replace size_t with int aren't correct header files. Have you ever seen a distribution in your life with multiple include trees because size_t was defined differently in one folder compared to the next?

An alternative is to NOT ship the header files at all. Ship a script that generates them appropriate to the compiler switches being used. After all, if you don't consider those files editable by the user after they are generated, why ship them?

**Comment by WalterBright** [comment]
> Walter makes the argument that Stefan's solution opens a floodgate, where the problem applies to all aliases

Oh, I guarantee the floodgates of bug reports will come in. "Why doesn't my alias get reflected in the .h file? D is stupid." Having magic case after magic case leaves us in a very poor position.

**Comment by WalterBright** [comment]
> Reopening because utilising originalType did not work, it's null in certain cases.

Look into fixing it for those cases.

**Comment by Geod24** [comment]
> An alternative is to NOT ship the header files at all.

Sometimes it's not possible. Be it a company policy, difficult CI, poor build system support, complication of the release process, or simply for the compiler's ability to bootstrap. Having both options available is definitely a must.

As for the rest, strong agreement. `originalType` is the way to go ATM.

**Comment by UplinkCoder** [comment]
My proposal is for size_t to not be an alias. But a builtin type. Exactly
because special casing the cpphdrgen to detect the size_t alias and
treating it differently to any other alias is madness.

**Comment by UplinkCoder** [comment]
Also this pr is not to be taken as is, I will get rid of the mangle
differences it currently generates.

**Comment by GooberMan** [comment]
> An alternative is to NOT ship the header files at all. Ship a script that generates them appropriate to the compiler switches being used.

It's an alternative, sure, but I feel time would be better spent by suggesting _good_ alternatives.

Source: Binderoo does this, but Binderoo isn't trying to generate distribution-quality library headers.

**Comment by UplinkCoder** [comment]
Look at the code which is required to work around the lack of `size_t`
https://github.com/dlang/dmd/compare/master...UplinkCoder:size_t_dtoh?expand=1

**Comment by TurkeyMan** [comment]
> An alternative is to NOT ship the header files at all. Ship a script that generates them appropriate to the compiler switches being used.

The whole point of a binary distribution is that the client does not have the source, and/or does not need the capability of being able to compile the source. This is especially important for D, where using it to produce a binary lib with headers doesn't require that a thousand engineers all must have D tools on their workstations, and that we work out how to plug D workflow into XCode, Android Studio, etc.

**Comment by TurkeyMan** [comment]
> Oh, I guarantee the floodgates of bug reports will come in. "Why doesn't my alias get reflected in the .h file? D is stupid." Having magic case after magic case leaves us in a very poor position.

That's absurd; this here is absolutely not "magic case after magic case", as I demonstrated above, `size_t` is the 3rd most commonly used type that exists... it pervades EVERYTHING EVERYWHERE.
Nobody will say "size_t works, why doesn't my alias?", because I suspect most people would not imagine that `size_t` is just an alias even for a second. And if it's not an alias, then you short-circuit your hypothetical above.

My motivating case today is not isolated, it's just another instance of this issue which emerges periodically for as long as I've been using D. It just so happens that this one is more important to address than previous cases, where some awkward workaround might have been acceptable.
Workarounds are always bad though; they add complexity where it shouldn't exist. I think there's a very strong case that `size_t` specifically should be special; because it really honestly is. It's the 3rd most common type to appear anywhere, and there are many cases of code that need to identify that a value is size_t.

The hack with `origType` may solve our issue here, and I'll be happy to resolve this issue in whichever way, but this issue is periodic, and I reckon we should fix it proper rather than with brittle hax...

**Comment by Geod24** [comment]
> I think there's a very strong case that `size_t` specifically should be special; because it really honestly is

I don't see it. What make it so special that it can't be an `alias` ?

> and I reckon we should fix it proper rather than with brittle hax...

This seems to be the only point we agree on. And the proper fix is *not* to make `size_t` special, it is to improve the way aliases are handled.

**Comment by UplinkCoder** [comment]
I reckon we should do both!
remove the size_t alias status, AND improve the way aliases are handled.

Am Mi., 26. Feb. 2020 um 20:00 Uhr schrieb Mathias LANG <
notifications@github.com>:

**Comment by WalterBright** [comment]
> this here is absolutely not "magic case after magic case"

After dealing with bug reports for 30 years ago, I guarantee it will be magic case after magic case. I also guarantee you're still going to have problems with `size_t`, such as the following:
```
static if (size_t.sizeof == 4) { ... } else { ... }
```
It just ain't going to be robust.

I still recommend going with `originalType` if that can work. Alternatively, you can do things like:
```
struct size_t { .size_t value; alias value this }
void foo(size_t x);
```
and it will come out in the .h file as type `size_t`. You can hide the `struct size_t` in an import or something that is only seen when generating the `.h` files. The beauty of this is you can use it for whatever other types you need it for, and you can use it *right now*.

**Comment by TurkeyMan** [comment]
> After dealing with bug reports for 30 years ago, I guarantee it will be magic case after magic case

This feels like a straw man though, there's always the next issue... I can't imagine `size_t` being a primitive would lead to people complaining about aliases not working; and I suspect that fear actually becomes _less_ likely if it was a primitive. These are different topics.

> `static if (size_t.sizeof == 4) { ... } else { ... }`

Why would code like that change from how it is today? I don't think this discussion has any affect on code like that.

> The beauty of this is you can use it for whatever other types you need it for, and you can use it right now.

I understand hacks like this, and I've done plenty of workarounds for my own code. The reason it feels unsatisfying for `size_t` in particular is because of the things I've said; it's EVERYWHERE. It's not just some type from some library that I chose to use in my code; it emerges from core language and core libraries everywhere.
So, it's not that "I decided to use `size_t` in my library, so I'll deploy this hack"; `size_t` just appears all the time, because it's fundamental. It emerges all on its own, and it's annoying to handle in all those places.

If there's code generation, or functions synthesised by copying parameter lists from other functions, these sorts of things often suffer from the signal loss. The origType solution might work... but it makes me much more nervous, it seems like another bandaid.

**Comment by UplinkCoder** [comment]
@WalterBright I'd like your opinion on the `sizeTSwtichCase` mixin.
It does what is needed but I am weary of introducing mixins into dmd.

**Comment by WalterBright** [comment]
> it's annoying to handle in all those places.

```
import manu : size_t;
```
```
module manu;
version (GenerateHFile)
    struct size_t { ... }
else
    alias size_t = Object.size_t;
```
and you're done. You can also redirect dmd to read a different object.d when generating .h files that has the `struct size_t` in it. This shouldn't be a big deal for you, as you've already stated that you are going to be generating the .h files, not the customer.

> it seems like another bandaid.

It's always going to be a bandaid. Picking half of semantic analysis, ignoring the other half and expecting it to work 100% just leads to frustration.

**Comment by TurkeyMan** [comment]
```
version (GenerateHFile)
    struct size_t { ... }
else
    alias size_t = Object.size_t;
```
Compiling with -HC doesn't not compile your code. The entire point of this effort, and the reason it's integrated into the compiler and not separate build tooling is that it's a one-pass solution. If you're iterating on code, you're emitting binaries. Substituting `size_t` for something else is not likely to compile seamlessly... because it's _everywhere_, and it pervades everything.

Build machines might get away with a multi-pass hack like that, but developers doing their job are why we're here.

**Comment by UplinkCoder** [comment]
@WalterBright `size_t` is not half baked. it is supposed to go trough all of semantic and come out a distinct type with all the properties of the integral type which matches it's size.

**Comment by WalterBright** [comment]
> size_t is not half baked.

It is for the .h generation for the simple reason of code like this:
```
int x = cast(int)size_t.sizeof;
```
Now, what are you going to do when `mixin` is using the value of x in its computations generating declarations to be emitted to the .h file? You might say "don't write code like that", but people inevitably will and their .h file will be silently broken and I'll be stuck with the bug reports.

> Compiling with -HC doesn't not compile your code.

It does because you're relying on `mixin` which *compiles* and *executes* code at compile time.

What you're asking for is fundamentally unsound. The solution I suggested leaves the unsoundness out of the language and at the discretion of the users. It *exposes* the unsoundness so it can be accounted for, rather than *hiding* it.

**Comment by TurkeyMan** [comment]
> > Compiling with -HC doesn't not compile your code.

> It does because you're relying on mixin which compiles and executes code at compile time.

I don't understand this sentence... I don't recall talking about `mixin`-ing? And I'm also not sure what you're saying...
I was trying to say your use of version doesn't work, because code that uses an alternative `size_t` probably won't actually compile (against all the other code that uses normal `size_t`)...
No such `version` exists. There aren't 2 'modes'.

> What you're asking for is fundamentally unsound.

What's unsound about `size_t` being a thing? Is `uint` unsound? You haven't explained why it would be bad for `size_t` to be a type, and I don't understand the trouble. What I do understand is the trouble of it not being a type. I am periodically reminded.

> The solution I suggested leaves the unsoundness out of the language and at the discretion of the users. It exposes the unsoundness so it can be accounted for, rather than hiding it.

Users can't re-define `size_t`, it's a fundamental language type... I don't think you can redefine `size_t` locally to your module and expect your code to compile normally and work.

**Comment by WalterBright** [comment]
> I don't recall talking about mixin-ing?

https://github.com/dlang/dmd/pull/10828#issuecomment-590734697

> Users can't re-define size_t

They can change `size_t` properties with -m32 and -m64 switches. And you  use mixins, which can use properties, and hence generating .h files with -m32 and using them with -m64 is *unsound* with regards to `size_t`.

Use:
```
long i = size_t.sizeof;
```
in your mixin computation and you're hosed.

**Comment by UplinkCoder** [comment]
@WalterBright having a separate `size_t` type would allow you to catch that mistake at compile-time if you wanted.

**Comment by WalterBright** [comment]
It's not a mistake to use `size_t.sizeof`. It's the *point* of having `.sizeof`. With grep I find 11 uses of this in dmd, none of them are bugs. I find a screenfull in Phobos. Several screenfulls in druntime. And that's not counting indirect such `sizeof` uses via aliases.

Baking `size_t` as a basic type into the language creates an *insoluble* problem w.r.t. generating .h files that are supposed to work with both `-m32` and `-m64`. Such must not be baked into the core language.

**Comment by UplinkCoder** [comment]
okay what about `real.sizeof` then?

**Comment by TurkeyMan** [comment]
> Use:
> `long i = size_t.sizeof;`
> in your mixin computation and you're hosed.

True. I don't think I've ever seen or written code anything like that though. I use `size_t` because I _don't_ want to coerce the size type to a hard integer type, not because I do. Use of `size_t` tends to be mutually exclusive with code like you just showed.

**Comment by WalterBright** [comment]
> I don't think I've ever seen or written code anything like that though.

`grep size_t.sizeof *.d` and you'll find plenty of uses of it in dmd, phobos, and druntime. They're not bugs, either.

**Comment by UplinkCoder** [comment]
That is because you deal with storing pointers. And it's not done in meta
code.

**Comment by TurkeyMan** [comment]
> `grep size_t.sizeof *.d` and you'll find plenty of uses of it in dmd, phobos, and druntime. They're not bugs, either.

That doesn't mean they influence function signatures though. Of course implementation code encounters target-dependent logic.

**Comment by WalterBright** [comment]
> That doesn't mean they influence function signatures though.

Again, you brought up your extensive use of mixins. But I understand, you're saying that in your application you don't do this. I have to think in broader terms, however.

The most sensible thing for your application is release two sets of .h files, one for -m32 and one for -m64, along with the two corresponding two sets of object libraries you're dealing with anyway. Simple. Done. No special cases. No bugs.

(Note that there is no `size_t` type in C or C++, either.)

**Comment by WalterBright** [comment]
BTW, it occurs to me that the only difference between the -m32 and -m64 headers is likely to be `uint` in one and `ulong` in the other, in the same places. This suggests a simple script could compare the two files line by line and replace such pairs with `size_t`.

The script could also let you know if other differences exist, which would be something you'd probably like to investigate.

**Comment by TurkeyMan** [comment]
> Again, you brought up your extensive use of mixins. But I understand, you're saying that in your application you don't do this. I have to think in broader terms, however.

No. I'm really lost... Yes we use mixins extensively. As always, reflection and code generation are usually the primary reason D may be considered over C++.
What I said is, code that does something like `static if (size_t.sizeof == 8)` doesn't appear in places that would affect function signatures. Code like that might appear inside a function when calculating allocation sizes or something, but logic that makes distinction on the size of `size_t` is mutually exclusive with using `size_t` in function signatures; the point of `size_t` in an API is self-defeated if you have target specific logic applying to the API that way.

> (Note that there is no size_t type in C or C++, either.)

That's not quite true... and I'm not here to argue we should copy C++ mistakes. In C++, `size_t` is a pseudo-magic type in the compiler which implements different implicit cast rules.
```c++
size_t x;
#if x64
uint64_t a = x; // no problem
uint32_t b = x; // ERROR: truncation
#else
static_assert(sizeof(size_t) == sizeof(uint32_t)); // it's a uint
uint64_t a = x; no problem
uint32_t b = x; // WARNING: portability warning; size_t assigned to 32bit integer!
                // actually an ERROR, because everyone uses warnings-as-errors.
#endif
```
The compilers obviously implement `size_t` as an appropriately sized int, but there is special sauce in the semantics where implicit conversion rules are applied; since they MAY be 64bit types, any code that assumes they're <64 bits is not portable, which defeats the purpose of using `size_t` in the first place. Implicit conversions like I show above produce portability warnings, rather than working silently.

I don't think you've told us any reason yet why `size_t` as a primitive type would cause problems. As an alias, it periodically causes is various kinds of problems... I'd like to know how being acknowledged as a real thing could be more trouble than it causes us today. This topic on trial is not the only case of problems, this class of issue comes up fairly often in various situations.
D has reflection, and code generation, and the information loss is material, much more so than C++ where the information loss can't really do much to affect your code anyway.

**Comment by TurkeyMan** [comment]
Actually, it hadn't occurred to me before, but the C++ semantic I described above will cause us 32bit compile errors if `size_t` is lost from header files.
C++ calling code which has a `size_t` can't actually call a function where `size_t` was replaced with `uint32_t`... it will emit the portability warning (error) above :/
I mean, I intuitively knew this was a problem, and I think that's actually why I'm making this whole noise in the first place. I just didn't think of the concrete issue until I spelled it out ;)

**Comment by WalterBright** [comment]
> code that does something like static if (size_t.sizeof == 8) doesn't appear in places that would affect function signatures. 

When using mixins, you're using CTFE, which is running functions at compile time. `.sizeof` isn't applied just to `size_t`, it's applied to fields and structs that may declare members with `size_t`. Even if you, yourself, never use `.sizeof` on anything, making it a type opens up an *unfixable* special case in the language that will *silently* produce *wrong answers* for anyone using it.

It's one thing to not realize this could happen, it's another to *knowingly* wire this into the language.

> That's not quite true.

It is true. On my Linux system, for example, it's:
```
typedef long unsigned int size_t;
```
I.e. not a type but an alias to a type, and the compiler has no knowledge of it. It is not special.

The solutions I showed you are simple and will work for your use case. They also suffer from the `.sizeof` problem, but they won't be a core language problem. It'll be up to the user to deal with it. I.e. it's one thing for the user to use an unreliable hack at their own risk, it is quite another to wire it into the language and pretend it is robust and encourage people to use it.

Given the straightforward suggestion https://github.com/dlang/dmd/pull/10828#issuecomment-592304868 I can't justify moving ahead with hacking the compiler.

**Comment by GooberMan** [comment]
> It is true. On my Linux system, for example, it's:
> 
> ```
> typedef long unsigned int size_t;
> ```

And in the Windows SDK, it is:
```
typedef unsigned __int64 size_t;
```

This built-in type native to MS compilers generates a different symbol for mangling than unsigned long int. This point is the entire reason why size_t as a basic type is attempting to be introduced, and why this thread has gone on for so long dancing around a bunch of unrelated issues like mixins and CTFE.

Perhaps the "correct" solution so to speak would be to correctly support the native __int64 types and alias size_t to them? It would still generate funky header files where a user would reasonably expect to see size_t instead of a compiler-specific unaliased type, but it would be one step closer to "correct".

**Comment by don-clugston-sociomantic** [comment]
I am excited by the concept of making `size_t` a built-in type. Not so much because it is a type that can vary in size (I think that 32-bit machines won't have any commercial relevancy for much longer) but because they are so ubiquitous. `size_t` is a really magical type since it's created by `arr.length` and this makes it a much more important type than it is in C++.


Almost all use of unsigned types in D are because size_t is unsigned. I have always maintained that unsigned size_t is a great evil. It really should have been a signed type. size_t is always a positive number, it's not "unsigned". You never want size_t to have `modulo 2^^64` semantics, but that's what you get.
The fact that it is unsigned means that any subtraction involving a size_t is potentially problematic.

eg.  auto x = p.length - q.length;   // if q is longer than p, this will be > long.max!!!

This isn't true of genuine unsigned values. Eg if you have two hashes, and you subtract them, it's perfectly ok that they wrap.

It would be wonderful if we could change it to have the semantics that people mistakenly expect it to have: an integer >= 0 (ie, the common subset of long and ulong). And then tell people to stop using unsigned types for that.

So `size_t` in code has *two* magical properties: (1) the obvious fact that it changes based on the architecture; and (2) the desired semantics are slightly different.

This is not true for any other alias.

**Comment by WalterBright** [comment]
> Perhaps the "correct" solution so to speak would be to correctly support the native __int64 types

D already does as `__c_ulonglong`. Note that `size_t` is still *not* a native VC++ type.

> The fact that it is unsigned means that any subtraction involving a size_t is potentially problematic.

There's no getting around the fact that fixed integer arithmetic, signed or not, is potentially problematic. There is no magic solution, not even Java's avoidance of having any unsigned types at all. It's like floating point. All the "solutions" I've seen simply caused other problems.

**Comment by WalterBright** [comment]
Let me amplify that a bit. Dealing with complexity with a bad abstraction is worse than no abstraction. The latter is left up to the programmer, and he can make it work right.

An example of a bad abstraction is string autodecoding. It seductively kinda sorta works, but when it doesn't, you're faced with a lot of bad alternatives that are much worse than no autodecoding at all.

Not to stray too off topic, but DIP1027 is another example of a very simple construct, that has some minor edge cases. But nobody likes that, they want a complex abstraction that has problems that are much more difficult to deal with. (No, I don't want to argue about that here, see the n.g. thread for that.)

`size_t` is a very old and very well known abstraction, with well known problems. As I've repeatedly pointed out in this thread, building it in as a magic type fixes one problem and creates a laundry list of new, unfamiliar, unexpected, and undetectable problems. I've given Manu a straightforward and simple fix for his specific problem. We should stop wasting time on a fruitless pursuit of false perfection.

**Comment by don-clugston-sociomantic** [comment]
> > The fact that it is unsigned means that any subtraction involving a size_t is potentially problematic.

> There's no getting around the fact that fixed integer arithmetic, signed or not, is potentially problematic.

No, that's a false analogy. By "potentially problematic" I mean that it is ALWAYS a bug.
You never, ever, ever want unsigned arithmetic when subtracting a `size_t`. You never want wrapping semantics.
The only time when it isn't a bug is when it is inside an `if` statement that checks it explicitly, so that unsigned arithmetic is the same as signed arithmetic.
You always want a signed arithmetic result. 

>  There is no magic solution, not even Java's avoidance of having any unsigned types at all. It's like floating point. All the "solutions" I've seen simply caused other problems.

That's an association fallacy, Walter. This isn't the general problem of overflow which indeed doesn't have a good solution, it's a simple type error in the language fundamentals.. Lengths of arrays have a sign, and that sign is positive. Making them unsigned was one of the biggest own goals in D1.

Like you, I used 8 bit and 16 bit machines where it was common for arrays to take up more than half of the address space. Those days are long gone and D never supported such systems anyway.
It's quite ridiculous that we are still paying the price for that. Note though that even on those systems, the typing rules were still not correct. When such a value is cast to a larger value, it should be zero-extended not sign-extended (so unsigned types give the correct bit pattern) but the resulting value should be a signed type.

Note that I'm not passing judgement on this PR specifically. I'm just pointing out that `size_t` is incorrectly typed.
Now *maybe* this could also be fixed with value range propagation. But it really is a genuine problem. Whenever anyone talks about "signed-unsigned mismatch" the actual underlying problem is almost always the wrong type of size_t.

**Comment by WalterBright** [comment]
There are still programs that *do* need an unsigned `size_t`, specifically 32 bit programs. These haven't gone away yet.

Secondly, making `size_t` signed and trying to match it up with unsigned `size_t` in C/C++ is something I'd expect seeing problems from.

**Comment by don-clugston-sociomantic** [comment]
> There are still programs that do need an unsigned size_t, specifically 32 bit programs.

Do you have evidence for that? I personally don't believe there 's anything that actually requires unsigned `size_t`. In my 32 bit code I have always wanted signed size_t.

Think about it --- the case where it makes a difference is where it spans more than half the address space. That's possible only when you have an array of bytes. And you can only have one such array.
If this is the case then multiplication even by 2 gives an overflow, so that's not a problem.

**Comment by don-clugston-sociomantic** [comment]
It would be really interesting to make `size_t` signed in a PR, and see how much breaks :)

**Comment by UplinkCoder** [comment]
At the present moment introducing a basic type without a constant mangled name is a lot of work and requires deep changes to how dmd handles types.

-- closing

**Comment by WalterBright** [comment]
> the case where it makes a difference is where it spans more than half the address space

This started happening when file sizes got big (and memory mapped files). Lots of programs would fail with this. I've also noticed that Microsoft Moviemaker starts behaving erratically when the movie is bigger than 2G.

> It would be really interesting to make size_t signed in a PR, and see how much breaks

The problem is much of the breakage would be silent. You wouldn't know.

**Comment by don-clugston-sociomantic** [comment]
>> the case where it makes a difference is where it spans more than half the address space
> This started happening when file sizes got big (and memory mapped files). Lots of programs would fail with this. I've also noticed that Microsoft Moviemaker starts behaving erratically when the movie is bigger than 2G.

I don't think that's related, just sounds like a bug to me. Any code that uses a size_t for file size is broken! If you use unsigned, you'll get problems once you get to 4Gb, it doesn't make the code any more correct. And even back in the days of MS-DOS 1.0, file sizes were much bigger than 16-bit size_t so it's not a mistake anyone should have ever made.

>>    It would be really interesting to make size_t signed in a PR, and see how much breaks
> The problem is much of the breakage would be silent. You wouldn't know.

True, some would be silent breakage, but the amount of compilation errors and unit tests failures should give a pretty good estimate. It could potentially refute my claim that hardly anything would be affected. 
Stefan did try it, and it showed one place in druntime that didn't compile (related to debug info). That was surprising to me, I wasn't expecting compilation errors.
OTOH a large fraction of all of the silent bugs I've found in my own D code, were caused by `size_t` being unsigned. In some cases these bugs lay dormant in production code for years before being triggered. It's genuinely a big problem.

**Comment by WalterBright** [comment]
> Any code that uses a size_t for file size is broken!

Lots of programs use memory mapped files (size_t) and simply load the file into one giant array. This is why they broke when files got bigger and bigger.

> a large fraction of all of the silent bugs I've found in my own D code, were caused by size_t being unsigned

That hasn't been my experience, likely because I've had so much experience with 16 bit size_t.

> some would be silent breakage

This is just trading one set of well-known problems to another set of unknown problems. Personally I'd have problems with it being signed, because it being unsigned is so well-ingrained into me. Then there's the issue of moving back and forth between D and C/C++ and getting discombobulated about size_t. I often slip and make mistakes because D works one way and C/C++ another.

**Pull request closed without merging**